Extend TestingGuide (adds notes on formatting)

banach-space opened the pull request.

joker-eph commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+* Follow existing conventions.
+
+By applying these best practices, we leverage available tools (e.g., test
+function names) to make tests easier to discover and maintain.
```
This a good addition: I would still add some documentation in comment to a test: it's not always clear just from a test name what's going on, and the same logic as in code applies (document the why, the context, the "what's no obvious from the code").

banach-space replied:
Yes, comments are also super important, thanks for raising this!

I have some ideas on "comments", so let me treat this as encouragement to extend this PR with a dedicated section on documentation :)

joker-eph reviewed (commented):
Thanks!

jpienaar commented on website/content/getting_started/TestingGuide.md:
````diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+
+  return %ld : vector<16xf32>
+}
+```
````
Nit: newline after block. I think our renderer is particular about it.

jpienaar reviewed (commented):
Thanks! Yes discoverability is important here.

rengolin commented:
Awesome addition. All good practices that we've been telling each other for years but always too lazy to encode in a document. :)

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+
+When adding new tests, strive to follow these two key rules:
+
+1. **Follow the existing naming and whitespace style.**
```
When you say 'existing style', what scope do you have in mind?

banach-space replied:
I’m deliberately avoiding specifying this :)

Making things globally consistent would be great for... consistency. However, that assumes **one size fits all**, which isn't always realistic. Given the current state of things, I also feel this would be an impractical goal.
The same applies to **per-file** vs **per-directory** consistency. For example, "mlir/test/Dialect/Vector" has around 50 files - enforcing 100% consistency across them would require significant effort and churn.

My view is that contributors and reviewers should make the right call on a case-by-case basis while being mindful of:

* Reinventing the wheel in terms of formatting.
* New tests ignoring existing conventions (whether at the file or directory level).

Ultimately, I see this as a **gradual** process. Let’s start with these guidelines to help "correct the course." Over time, we can refine them - possibly transitioning from **guideline** to requirement if it makes sense.

WDYT?

kuhar replied:
I've worked on a few different codebases that approached this differently: some preferred local consistency, some global (e.g., google3). I agree that this doesn't have to one or the other and there's some room for interpretation, but at the same time I'd like to have some codified incentive that will push us towards **higher** quality tests/code.

For example, we could say we prefer global consistency unless the local standards are either higher or more specialized. If a PR adheres to the global standards but sticks like a sore thumb compared to other local code, I'd think that in most cases sending a separate clean up PR would be a reasonable request.

banach-space replied:
Thanks, Jakub! Sorry for the delayed reply—I haven’t had much time for this PR today. 😅  

My overall sentiment is that we should approach defining these guidelines as an **incremental process**, avoiding overly strict or prescriptive rules.  

> For example, we could say we prefer global consistency unless the local standards are either higher or more specialized.

I personally like that, but I worry that others may disagree, and we might get stuck discussing it. Also, if we ask for **global consistency**, who is actually going to help us get there with what we have today?  

We shouldn't require new contributions to be **fully consistent** with everything we already have, given how... inconsistent our tests are. 😅  

---

### **Proposed Levels of Consistency**
How about defining **different levels of consistency**?  

1. **Per-PR consistency**  
   - When adding new Ops, documentation, and tests, they should use an identical style.  
   - **Mandatory**.  

2. **Per-file consistency**  
   - If a file is already consistent, maintain that.  
   - If a file is inconsistent, create a GitHub issue to address it.  
   - **Mandatory**.  

3. **Per-directory consistency**  
   - This might require a lot of churn, so I’m unsure about enforcing it strictly.  
   - **Nice-to-have**.  

4. **Per-project (MLIR-wide) consistency**  
   - Achieving this would require **volunteers** to refactor existing tests.  
   - **Nice-to-have**.  

---

This is just a rough idea - suggestions are welcome!

kuhar replied:
> My overall sentiment is that we should approach defining these guidelines as an incremental process, avoiding overly strict or prescriptive rules.
> I personally like that, but I worry that others may disagree, and we might get stuck discussing it. 

This is a very good point, let's revisit this once the initial version lands.

> Also, if we ask for global consistency, who is actually going to help us get there with what we have today?
> We shouldn't require new contributions to be fully consistent with everything we already have, given how... inconsistent our tests are.

Maybe 'consistency' is not the best term here. The idea would be that we would be getting closer to the agreed on 'testing guidelines' (similar to the [coding standards](https://llvm.org/docs/CodingStandards.html)) incrementally as we land new code, and refactor old tests along the way. Having the coding standards codified and linkable is a fantastic tool in code reviews because it allows you to point out the best practices without making it feel like you are making one-of arguments against someone's implementation. I'd hope that with testing guidelines we can accomplish something similar and gradually steer tests towards higher maintainability/quality/readability.

kuhar replied:
IE, my proposed rule of thumb would be something along the lines: 'Follow the testing guidelines while adhering to the local naming and formatting style.'

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+  return %ld : vector<16xi8>
+}
+
+// CHECK-LABEL:   func @maskedload_regression_3(
+// CHECK-SAME:        %[[A0:.*]]: memref<16xf32>,
+// CHECK-SAME:        %[[A1:.*]]: vector<16xf32>) -> vector<16xf32> {
+//      CHECK:    return %[[A1]] : vector<16xf32>
+func.func @maskedload_regression_3(%arg0: memref<16xf32>, %arg1: vector<16xf32>) -> vector<16xf32> {
+  %c0 = arith.constant 0 : index
+  %vec_i1 = vector.constant_vec_i1 [0] : vector<16xi1>
+
+  %ld = vector.maskedload %arg0[%c0], %vec_i1, %arg1
+    : memref<16xf32>, vector<16xi1>, vector<16xf32> into vector<16xf32>
+
+  return %ld : vector<16xf32>
+}
```
Do we need all 3 examples to drive your point or could we limit this to one or two?

banach-space replied:
"Generalization Takes Three Examples" :)

It allows me to go into a bit of nuance in this paragraph (i.e. to identify the commonalities):

```mlir
  ##### Step 2: Improve Test Naming

  Instead of using "regression" (which doesn't add unique information), rename
  tests based on key attributes:

  * All examples test the `vector.maskedload` to `vector.load` lowering.
  * The first test uses a _dynamically_ shaped `memref`, while the others use
    _static_ shapes.
  * The mask in the first two examples is "all true" (`vector.constant_mas
    [16]`), while it is "all false" (`vector.constant_mask [0]`) in the thir
    example.
  * The first and the third tests use `i32` elements, whereas the second uses
    `i8`.
```

That's not something I feel super strongly about. Usually its tricky to see the "optimal" solution on first iteration and I guess that's the case here.

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+
+* `@{negative_}?maskedload_to_load_{static|dynamic}_{i32|i8}_{all_true|all_false|mixed}`.
+
+#### Final Points - key principles
```
```suggestion
#### Final Points - Key Principles
```

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+* The mask in the first two examples is "all true" (`vector.constant_mas
+  [16]`), while it is "all false" (`vector.constant_mask [0]`) in the thir
+  example.
+* The first and the third tests use `i32` elements, whereas the second uses
```
The key attributes should reflect what's important for the tested transform, in this case having different data types is good for completeness but they don't trigger any special/separate logic.
It would be sufficient to highlight one case that uses different data type with, for example, a suffix and omit data type for others.

I know that's not the intention behind this example but it can be misused to push for "enterprisification" (classic java gem: https://projects.haykranen.nl/java/) of naming with combinatorial explosion of all the little details.

banach-space replied:
Added a short paragraph in this [commit](https://github.com/llvm/mlir-www/pull/216/commits/6f7b7697ac7911f935fa9db047a4f6e594b6fe06). Let me know what you think - do we need more?

adam-smnk replied:
Great clause 👍
It conveys the spirit of these guidelines and defends against too pedantic rule lawyers.

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+
+When adding new tests, strive to follow these two key rules:
+
+1. **Follow the existing naming and whitespace style.**
```
nit: Should the existing bad style (test_1, test_2, ...) prevent naming new test cases more reasonably?

umangyadav commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,248 @@ func.func @simple_constant() -> (i32, i32) {
+* The first test uses a _dynamically_ shaped `memref`, while the others use
+  _static_ shapes.
+* The mask in the first two examples is "all true" (`vector.constant_mas
+  [16]`), while it is "all false" (`vector.constant_mask [0]`) in the thir
```
```suggestion
  [16]`), while it is "all false" (`vector.constant_mask [0]`) in the the
```

dcaballe commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,463 @@ func.func @simple_constant() -> (i32, i32) {
+* All examples test the `vector.maskedload` to `vector.load` lowering.
+* The first test uses a _dynamically_ shaped `memref`, while the others use
+  _static_ shapes.
+* The mask in the first two examples is "all true" (`vector.constant_mas
```
`mas` -> `mask`

dcaballe commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,463 @@ func.func @simple_constant() -> (i32, i32) {
+* Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.
+* Use high-level block comments to describe patterns being tested.
+* Think about maintainability - comments should help future developers
+	understand tests at a glance.
```
weird indentation?

dcaballe commented:
Thanks a lot! We definitely needed some specific guidelines here. Thanks for putting in the time! 

I think it might be helpful to clarify the level of enforcement we expect for these guidelines and the degree of flexibility we should allow. I fully agree with the motivation points highlighted in the PR but I'm also worried that we make things painfully rigid. If we set the bar too high, there’s a risk that testing becomes tedious enough to potentially impact actual coverage. I’m also worried that some of the renaming/formatting aspects of the guidelines take the focus away in the code reviews from other aspects that should have more attention.  

To alleviate some of this, esp. if we are going after a high level of enforcement, I think we should put some effort into automating some of the renaming/formatting aspects. For example, we could extend `generate-test-checks.py` (or implement a new tool) for MLIR so that it generates more meaningful CHECK variable names by default, derived from the dialect or op name, for example. That could greatly enhance consistency by default while minimizing manual effort.

I also feel consistency is open to interpretation. For example, if an existing test uses `V_INPUT_0` as a CHECK variable name and a new test is added with a `INPUT0` variable name, should we enforce a change? That name seems valid to me and we don’t enforce such a level of naming consistency on the C++ code. I think we may want to allow the same level of flexibility on the testing side. 

Anyways, some food for thoughts and probably unpopular opinions I wanted to share 😊. Great writeup and ideas overall. Thanks!

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+better, consider refactoring the file to adopt a single, consistent style —
+this helps improve our overall testing quality.
+
+This is also encouraged when the existing style leaves room for improvement.
```
nit: Specify `this` more explicitly. Does it refer to following the existing style or refactoring?

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+//       CHECK:      %[[TR:.*]] = vector.transpose %[[VEC]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
+//       CHECK:      vector.transfer_write
+
+/// The permutation map was replaced with vector.transpose
+// CHECK-NOT:       permutation_map
```
This example is also made by the second example. Could they be collapsed?

Also, while the previous 3 examples at least are driven by more detailed narrative, I feel these verbose examples are overkill for a message that is summarized with two bullet points (just to be clear - I mean smaller example is more readable, not adding more description ;) ).

Or perhaps the key parts could use some emphasis (bold etc.). Or the actual test details could be omitted for documentation clarity.
Blob of IR triggers tl;dr before I filter out the message.

adam-smnk replied:
Same for the convolution example.

banach-space replied:
Agreed, let me trim these examples. Let me know what you think.

> Or perhaps the key parts could use some emphasis (bold etc.).

Is it possible in code blocks in Markdown?

> Same for the convolution example.

Yes, it's quite long. It's a bit tricky to find good examples that are also short 🤔 But I totally agree that examples in this guide should be of highest possible quality. No pressure! 😅

adam-smnk replied:
> > Or perhaps the key parts could use some emphasis (bold etc.).
> 
> Is it possible in code blocks in Markdown?

Not certain but most likely no (in standard Markdown).
Docs and examples like these are always art of tradeoffs. Do these the examples need to be complete and self-contained? Or are they primarily a vehicle to guide (reader's gaze) through key concepts?

I like the cleanups for `Test Documentation Best Practices` section. Each code block has is a self contained example that showcases a concept with matching description.

In case of `Example: Improving Test Readability & Naming`, I feel it's hard to follow changes done with each step. Especially as each code block fills up my screen, I have to jump between current block, previous block, and the step description.

Since formatting is not really an option, maybe emphasis by omission of the common elements could help? Sth like:
Base case:
```mlir
// CHECK-LABEL:   func @maskedload_regression_1(
//  CHECK-SAME:       %[[A0:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[A1:.*]]: vector<16xf32>
func.func @maskedload_regression_1(%arg0: memref<?xf32>, %arg1: vector<16xf32>) {
  ...
  //       CHECK:   %[[LOAD:.*]] = vector.load %[[A0]][%[[C]]] : memref<?xf32>, vector<16xf32>
  %ld = vector.maskedload %arg0[%c0], %vec_i1, %arg1
  ...
}
```
Step 1 - Use Consistent Variable Names:
```mlir
// CHECK-LABEL:   func @maskedload_regression_1(
//  CHECK-SAME:       %[[BASE:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[PASS_THRU:.*]]: vector<16xf32>
func.func @maskedload_regression_1(%base: memref<?xf32>, %pass_thru: vector<16xf32>) {
  ...
  //       CHECK:   %[[LOAD:.*]] = vector.load %[[BASE]][%[[C]]] 
  %ld = vector.maskedload %base[%c0], %mask, %pass_thru
  ...
}
```
Step 2 - Improve Test Naming
```mlir
// CHECK-LABEL:   func @maskedload_to_load_dynamic_i32_all_true(
//  CHECK-SAME:       %[[BASE:.*]]: memref<?xf32>,
//  CHECK-SAME:       %[[PASS_THRU:.*]]: vector<16xf32>
func.func @maskedload_to_load_dynamic_i32_all_true(%base: memref<?xf32>, %pass_thru: vector<16xf32>) {
  ...
}
```
The doc section could open with IR blob with the original three test cases, show these reduced steps, and close with all three test cases shown again with all these steps applied.

As usual, different tradeoffs - just treat it as yet another possibility.
And of course, it is biased toward my style of reading where I prefer to do quick visual diff than read text 😉

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+While we generally document why, documenting what is valid and encouraged in
+cases where:
+
+* The he test generates long or complex output
```
```suggestion
* The test generates long or complex output
```

adam-smnk commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+Note that while the comments document what is happening (e.g., "Write the
+result back in one shot"), some variables — like `w` and `kw` — remain
+"enigmatic" and are not explicitly explained. This might leave us
+second-guessing their meaning at first. However, that’s intentional—their
```
nit: white spaces around dashes

banach-space replied:
I assume that you were referring to `intentional-their`?

`second-guessing` appears to be correct: https://dictionary.cambridge.org/dictionary/english/second-guess

Btw, English is my 2nd language -  I am always happy to be corrected :)

adam-smnk replied:
Oh yes, I mean the former. There was another one in a different paragraph that you already fixed with other changes.

javedabsar1 commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+2. **Consistently document the edge case being tested.**
+   - Clearly state what makes this test unique and how it complements other
+     similar tests.
+
```
4. **Names and Comments**
 Test names should reflect **what** is being tested. The **why** shouldnt be attempted to code in the name and unless obvious from code/context, must be added as comment but avoiding over-explainng..

banach-space replied:
These are great, thanks @javedabsar1 !  I'd like to keep this particular section very short (and limit the guidelines to two very high level points). Is it OK if I add these points in `#### What if there is no pre-existing style to follow?` instead?

javedabsar1 replied:
> These are great, thanks @javedabsar1 ! I'd like to keep this particular section very short (and limit the guidelines to two very high level points). Is it OK if I add these points in `#### What if there is no pre-existing style to follow?` instead?

Sounds good.

javedabsar1 commented:
Thanks @banach-space  for doing this work. much appreciated.

dcaballe commented:
Thanks, Andrzej. Yes, I mostly agree with all that you said! I think the example you provided above illustrates my specific concern:

>For a mild improvement, we could extend generate-test-checks.py to rename VAL_0 and VAL_1 to BITCAST_1 and BITCAST_2:
>
>// Automatically generated - possible near-future version (mild improvement)
>// CHECK-LABEL:   func.func @bitcast_2d(
>// CHECK-SAME:                          %[[ARG_0]]: vector<2x4xi32>) -> vector<2x16xi8> {
>// CHECK:           %[[BITCAST_1:.*]] = vector.bitcast %[[ARG_0]] : vector<2x4xi32> to vector<2x2xi64>
>// CHECK:           %[[BITCAST_2.:*]] = vector.bitcast %[[BITCAST_1]] : vector<2x2xi64> to vector<2x16xi8>
>// CHECK:           return %[[BITCAST_2]] : vector<2x16xi8>
>func.func @bitcast_2d(%arg0: vector<2x4xi32>) -> vector<2x16xi8> {
>  %0 = vector.bitcast %arg0 : vector<2x4xi32> to vector<2x2xi64>
>  %1 = vector.bitcast %0 : vector<2x2xi64> to vector<2x16xi8>
>  return %1 : vector<2x16xi8>
>}
>However, ideally, we’d want something context-aware, such as:
>
>// Human-curated, context-aware naming
>// CHECK-LABEL:   func.func @bitcast_2d(
>// CHECK-SAME:                          %[[ARG_0]]: vector<2x4xi32>) -> vector<2x16xi8> {
>// CHECK:           %[[UPCAST:.*]] = vector.bitcast %[[ARG_0]] : vector<2x4xi32> to vector<2x2xi64>
>// CHECK:           %[[DOWNCAST.:*]] = vector.bitcast %[[UPCAST]] : vector<2x2xi64> to vector<2x16xi8>
>// CHECK:           return %[[DOWNCAST]] : vector<2x16xi8>
>func.func @bitcast_2d(%arg0: vector<2x4xi32>) -> vector<2x16xi8> {
>  %0 = vector.bitcast %arg0 : vector<2x4xi32> to vector<2x2xi64>
>  %1 = vector.bitcast %0 : vector<2x2xi64> to vector<2x16xi8>
>  return %1 : vector<2x16xi8>
>}


Someone might find that `BITCAST_1` is perfect, while someone else might find that it’s not capturing enough information. Or the other way around! I’d personally argue that for a test with three `CHECK` rules, even `VAL_1` or `A` would be just fine. This is the kind of arguments that concerns me. As you well said, these guidelines are subject to interpretation and opinion so I think it's important that we somewhat define where the "acceptable band" is. 

Regarding the automation, honestly, I think having a tool that generates variables based on the dialect/op names would be a fantastic starting point and help define where the acceptable bar is. Of course, I'm not saying you should implement such a tool :)

banach-space commented:
> Someone might find that `BITCAST_1` is perfect, while someone else might find that it’s not capturing enough information. Or the other way around! I’d personally argue that for a test with three `CHECK` rules, even `VAL_1` or `A` would be just fine. 

I agree with you - for that small example, basic variable names would be totally fine (e.g. `VAL_1`). But note that that example was just to illustrate the limitations of `generate-test-checks.py`. When writing that, I had this much more involved example in mind:

https://github.com/llvm/llvm-project/blob/60cc3af0d93ecb8bfc9d6bebc6cbc395df3bb4b6/mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned-non-atomic.mlir#L85-L120

> This is the kind of arguments that concerns me. As you well said, these guidelines are subject to interpretation and opinion so I think it's important that we somewhat define where the "acceptable band" is.

I've added a note on "common sense" in this [commit](https://github.com/llvm/mlir-www/pull/216/commits/6f7b7697ac7911f935fa9db047a4f6e594b6fe06). Perhaps I can extend that to capture your feedback (which I agree with).

banach-space commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,473 @@ func.func @simple_constant() -> (i32, i32) {
+When adding new tests, strive to follow these two key rules:
+
+1. **Follow the existing naming and whitespace style.**
+   - This applies when modifying existing test files.
```
I like your suggestion, thanks! Shall we also ask people to create GitHub issues when a test file _does not_ follow a consistent style?

CC @kuhar (I mentioned something similar in another thread with you)

banach-space commented:
**Update 26/2/24**

The latest [commit](https://github.com/llvm/mlir-www/pull/216/commits/3aa153c6ff1c8ee1c6d02dd2544422a419f2c8fc) addresses comments from @javedabsar1 . Thanks Javed, great suggestions!

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+Test names should reflect _what_ is being tested, not _why_.
+
+Encoding _why_ in test names can lead to unnecessarily long and complex names.
+Instead, add inline comments where needed—but avoid over-explaining.
```
```suggestion
Instead, add inline comments where needed.
```
nit: I think this already implies our preference for conciseness.

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+inconsistent styles, try to identify the dominant one and follow it. Even
+better, consider refactoring the file to adopt a single, consistent style —
+this helps improve our overall testing quality. Alternatively, create a GitHub
+issue and leave a TODO in the test file taht you are modyfying.
```
```suggestion
issue and leave a TODO linking to it in the test file that you are modifying.
```

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+By applying these best practices, we leverage available tools (e.g., test
+function names) to make tests easier to discover and maintain.
+
+### Test Docmentation Best Practices
```
```suggestion
### Test Documentation Best Practices
```

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+A good rule of thumb is to think of yourself six months from now and ask:
+"What might be difficult to decipher without comments?"
+
+If you expect something to be tricky for future-you, it’s likely to be tricky
```
```suggestion
If you expect something to be tricky for the future-you, it’s likely to be tricky
```

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+
+* `@{negative_}?maskedload_to_load_{static|dynamic}_{i32|i8}_{all_true|all_false|mixed}`.
+
+#### What if there is no pre-existing style to follow?
```
I like this section, it aligns pretty close with what I had in mind in one of the comment threads. Thanks!

kuhar commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,490 @@ func.func @simple_constant() -> (i32, i32) {
+
+If the test file you are modifying lacks a clear style and instead has mixed,
+inconsistent styles, try to identify the dominant one and follow it. Even
+better, consider refactoring the file to adopt a single, consistent style —
```
I wonder if it would be worth spelling out that such refactoring should generally go to a separate PR, or if that's obvious.

banach-space replied:
This is something I'd leave under-specified.

I agree with you that such changes should be submitted as a separate PR. At least in general. However, there are downsides:
* more churn + more PR traffic.

Also, sometimes the noise level is fairly low and sending a dedicated PR might be too much. So, I'd trust our judgement on case-by-case basis. Is that OK? We can always refine later.

banach-space commented:
**(2nd) Update 26/2/24**

The latest [commit](https://github.com/llvm/mlir-www/pull/216/commits/79c6f22aa1e202a66a8dfdf9b006f7a6783f92b1) addresses the latest comments from @kuhar  . Thanks Jakub! 🙏🏻 

Also, apologies for taking my time responding to folks! There's a lot of great suggestion to go over - thank you!  I want to make sure that I carefully parse and incorporate all your great ideas :)

kuhar reviewed (approved):
LGTM

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
````diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+//      CHECK:   vector.transfer_write %[[RES_1]], %[[OUTPUT]][%[[C0]], %[[C0]], %[[C0]]]
+```
+
+Note that while the comments document what is happening (e.g., "Write the
````
"Though the comments document..."

Is a neater way to start.

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+
+Note that while the comments document what is happening (e.g., "Write the
+result back in one shot"), some variables — like `w` and `kw` — remain
+"enigmatic" and are not explicitly explained. This might leave us
```
"some variables — like `w` and `kw` — are not explained."

Simplify this. I know what you mean by explicitly, but you're about to talk about that in the next sentence anyway so no need to add more big words here :)

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+Note that while the comments document what is happening (e.g., "Write the
+result back in one shot"), some variables — like `w` and `kw` — remain
+"enigmatic" and are not explicitly explained. This might leave us
+second-guessing their meaning at first. However, that’s intentional - their
```
Cut the whole "This might leave us..." sentence. It just delays getting to the key point here from "However, that's...."

Also, we can skip the however.

"are not explained. This is intentional"

This makes the point clear and does not send the reader around the houses to get to it.

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+purpose becomes clear when studying the corresponding Linalg vectorizer
+implementation (or, when analysing how `linalg.conv_1d_nwc_wcf` works).
+
+It’s important to remember that comments should assist in understanding the
```
"Comments help you understand code, they do not replace the need to read it. Comments guide the reader, they do not repeat what the code already says."

Drop the important, mostly importantly.

If it wasn't important you wouldn't bother to write it down, plus the average new reader of this has no relative scale of importance so everything written is important anyway (which is good, it keeps people diligent).

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+to restate what the code already conveys.
+
+#### Final Takeaways
+* Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.
```
"Always document why, document what if you need to" ?

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+
+#### Final Takeaways
+* Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.
+* Use high-level block comments to describe patterns being tested.
```
"Use block comments to describe the patterns being tested"

Otherwise I'm wondering what a high-level block comment is, is this different from a low-level block, where are my basement level comments at?

I think I know what you mean, but if you can be more straightforward here, again, it's one less chance for confusion.

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+#### Final Takeaways
+* Prioritize documenting _why_ something happens, unless the _what_ is non-trivial.
+* Use high-level block comments to describe patterns being tested.
+* Think about maintainability - comments should help future developers
```
I would add "(which includes you)" here because we can't stress that enough.

Stressing it won't teach the lesson - only experience really does that, but it helps.

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+* Use high-level block comments to describe patterns being tested.
+* Think about maintainability - comments should help future developers
+  understand tests at a glance.
+* Comments should assist, not replace reading the code. Avoid over-explaining.
```
The last part seems like a bad transition, not quite a non-sequitur though. Make that into its own point?

* Avoid redundant explanations.

?

DavidSpickett replied:
This does feel a bit like:
* do a lot of this
* but don't do too much
* got it?

But this is inherent to the process, and will come down to reviewer preference so I think this is just a feature of communication in general.

banach-space replied:
It's very hard to find the right balance, yes. In fact, one of my goals is to hint that:

*  _Finding the right balance w.r.t. comments might be your most challenging task._

Let me rephrase this a bit to improve the flow. Perhaps I should re-write it altogether, but I need a bit more time to think about it. Alternatively, wait and see how people use these guidelines in practice 🤷🏻

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+
+##### Step 3: Add The Newly Identified Missing Case
+
+The update in Step 2 made us realize that we were missing an important edge
```
I don't think the we that includes me the reader realised anything here to be honest, you're about to tell me what you realised, the author of this.

Rephrase to avoid this.

"Step 2 made it possible to see that there is a case which is not tested."

DavidSpickett commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+
+
+#### Documenting the "What"
+While we generally document why, documenting what is valid and encouraged in
```
Also if you use "you" it's more direct like:
"You should always document why, but also..."

DavidSpickett reviewed (commented):
General comment: avoid as many contractions as you can.

This is part of some style guides but I know it's (!) not a universally liked rule. So I'm saying this because personally even I find it slightly clearer when contractions are not used and if you're writing a new document, the cost to remove them is low.

Otherwise, as someone who has spent 0 time in MLIR, I would find this document helpful. It'll have a lot of corner cases of its own but it gives me something to cite to justify my starting points.

banach-space commented:
**Update 28/2/24**

Incorporated most of the suggestions from @DavidSpickett , thanks! (see this [commit](https://github.com/llvm/mlir-www/pull/216/commits/8430470b36c3f801bed2fdc5a4670a683b3f428e)). I've "resolved" the corresponding threads and left the remaining ones for later.

banach-space commented on website/content/getting_started/TestingGuide.md:
```diff
@@ -372,3 +372,452 @@ func.func @simple_constant() -> (i32, i32) {
+The above approach is just an example. It may not fit your use case perfectly,
+so feel free to adapt it as needed. For example, for "folding" tests, it may
+make more sense to use "no" as prefix for negative tests (e.g.
+`@no_fold_<case>_<subcase>`). Key principles to follow:
```
Good point, let me remove it from here. I do want to preserve `@no_fold_<case>_<subcase>` as an example though, so I will _move_ it elsewhere (as opposed to _removing_ it altogether). 

For context, @adam-smnk and I discussed that very example in https://github.com/llvm/llvm-project/pull/122449 (in fact, that prompted me to create this PR). In general, I think that it's good to have additional examples to demonstrate various options. Mostly to avoid people think that there's only one correct way of naming things. But we obviously also want to avoid going too far.

banach-space commented:
**Update 2/3/25** 

Minor updates to address feedback from David: [commit](https://github.com/llvm/mlir-www/pull/216/commits)

javedabsar1 reviewed (approved):
LGTM. Thanks.

DavidSpickett commented:
Nothing was that misleading, so whatever you do with my remaining suggestions is fine with me.

banach-space commented:
**Update 5/3/25** 

This [commit](https://github.com/llvm/mlir-www/pull/216/commits/ddb155d94b6a669e29d3b4043a2146a0bce2b7eb) addresses the remaining comments from David, thank you!

banach-space commented:
Ping @joker-eph , @jpienaar and @umangyadav - would you like me to wait for more feedback from you? Or shall I ship it?

joker-eph reviewed (approved):
LG, thanks
